nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp

```diff
@@ -115,6 +116,13 @@ class InflationLayer : public Layer
     return cost;
   }
 
+  // Provide a typedef to ease future code maintenance
```

Comment:
I don't think you need either the object or the method, its already provided for you since you derive from the costmap layer class which has a costmap2D in the inheritance tree https://github.com/ros-planning/navigation2/blob/aaa97902c1e1bb9a509c4ee0d88b880afb528f20/nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp#L314

Reply:
I thought so first, but actually `InflationLayer` is not `Costmap2D` but just a `Layer`.
The other three layers (Obstacle, Voxel, Static) are derived from `CostmapLayer` which derived from `Layer` and `Costmap2D`. I also considered moving the mutex things from `Costmap2D` to `Layer` but they are used in `LayeredCostmap` which just uses `Costmap2D` as a private member. 

So I needed to redefine them.
One possible alternative is having mutex things in `LayeredCosmap` and `Layer` for future additional plugins.
What do you think?

Reply:
Huh got it, thanks. @AlexeyMerzlyakov maybe the costmap filter base class should just be a layer and not a costmap layer, then you won't have that pesky `costmap_` internally https://github.com/ros-planning/navigation2/pull/1882/files. I think where we left it is we'd have a `mask_costmap_` or similar so we could just get rid of that in general. 

Thanks for that note, I hadn't recognized that subtly.

Reply:
> maybe the costmap filter base class should just be a layer and not a costmap layer

Good point to have in `CostmapFilter` class from #1882 in order to avoid `costmap_` resource potential mishandle in the future. I've added locally some locks and it looks to work fine.
BTW, from first sight it looks like `CostmapLayer` class is not locking the its resources when making any work with them (e.g. during the `updateWithMax()`, `updateWithOverwrite()`, etc... routines). Why do not add the access lock into `CostmapLayer` instead?

Finally, this PR was merged, but I can not find any commit in the master/main branch related to. Am I wrong. or is it a some merge problem?

Reply:
You don't need the locking in update functions for costmap filters necessarily. Inflation needs it because it has an async callback function editing a cache table that is also used by update (that could be called at  the same time). I think how we've designed the costmap filters layer, there's no chance for that because we don't even allow it to update anything until after we've received something over the topic and set the variable.  We should though look over it again and verify that is true. The update with XYZ doesn't need a lock because we're updating the `master_grid` which is only edited by the plugins, in a single threaded order. The locks are to protect the local `costmap_` isntances that can be edited by multiple things if there are callback functions being processed in the active state (like inflation, static, voxel, obstacles)